### ISSUE TITLE
If GitHub used CRS, I wouldn't be able to post this bug

### ISSUE THREAD

**Issue by bdjnk** [SpiderLabs/owasp-modsecurity-crs#235]
Why you might ask? Because it contains the following markdown:

```
## testing

    function test() {
```

This can be tested at the [official demo page](http://www.modsecurity.org/crs-demo.html).

**Comment by csanders-git** [comment]
What exactly are you trying to show us?

**Comment by bdjnk** [comment]
ModSecurity with CRS seems to block the posting of code within markdown. This is a problem.

**Comment by csanders-git** [comment]
tl;dr: Well I agree that this is a false positive, I would say that by design this isn't a problem. The ModSecurity language includes the tools necessary to craft exceptions to existing rules such that can prevent false positives. This includes within the CRS ruleset.

The issue we are discussing deals with context. ModSecurity is designed as a defense and blocking anything falsely means it likely requires tuning. Unfortunately, due to its placement at the Webserver level, it doesn't understand the particular context of the pages themselves, this must be added by the maintainer. The problem being that there is no implicit way to know if 'function test() {' is the beginning of code injection (Which it could easily be in many cases) or a benign markdown. It errs on the side of detection, as to not do so would let attacks through.

This is extremely similar to saying ModSecurity blocks when I try and send www.test.com?comment=<html><body>test</body></html> ... Clearly in many situations you might want to send HTML code as an parameter, but because this often leads to HTML injection this is detected. There are however ways around this problem. In general CRS is designed to be very generic, addressing threats on a more specific basis (like many commercial rulesets do prevents such issues). Additionally, it is always preferred to customize rules to your site - whitelisting instead of blacklisting what is allowed.

Those solutions not withstanding CRS attempts to correct for similar problems. For example in the 3.x branch we allow users to configure which sites allow HTML checks, and by extension HTML through for things like comments. (https://github.com/SpiderLabs/owasp-modsecurity-crs/blob/v3.0.0-dev/modsecurity_crs_10_setup.conf.example#L420)

While we are open to ideas as to a particular remediation, I fear that modifying the check to allow your example through would also allow for vulnerabilities in different contexts to be let through. As you might have guessed, in general we don't remove whole rules because of a given false positive. Instead the ModSecurity language has support for modifying existing rules by adding exceptions. See https://www.trustwave.com/Resources/SpiderLabs-Blog/ModSecurity-Advanced-Topic-of-the-Week--(Updated)-Exception-Handling/

For example if we wanted to explicitly allow similar input containing code on the modsec demo site, we might add a rule similar to this (untested):
SecRule REQUEST_URI "@contains crs-demo.html" "id:1,chain"
   SecRule ARGS:Test "skip:1" # where the next rule would be the 950907 (RCE attempt)

You could also do things like:
SecRuleUpdateTargetById 950907 !ARGS:test

The latter would stop the rule from running on any ARGS names test. Granted its a little less specific. then the first.


In general doing this must be done on a case by case basis and opens a hole through ModSecurity which means that defenders should be weary. Defenders should fully understand the consequences of their actions by reviewing their code from a security standpoint and understanding that adding such an exception may leave them vulnerable. If they have properly used defense in depth then there will also be other protection mechanisms (such as proper coding techniques, IDS's, etc) that will prevent any attacks via that hole.

Hopefully this has given you some insight into how we deploy CRS in environments like github without breaking github or CRS. In the future please try to keep your issue titles and summaries more descriptive such that if necessary other people may find similar issues (and solutions) by using the search feature. I am going to close this issue but if you have other comments feel free to keep responding. Thanks

**Comment by bdjnk** [comment]
Thank you for the detailed and informative response.

It seems, and your reply confirms the impression I've received elsewhere, that ModSecurity often acts as server protection against security holes created by poorly designed web applications. I guess, from my perspective, the whole approach seems wrong.

**Comment by csanders-git** [comment]
To be clear it is the goal of CRS is to provide generic protections against classes of vulnerabilities, not ModSecurity. ModSecurity itself just allows the maintainer control to filter on whatever they want, such as by generating whitelisting. In general CRS is setup to attempt to defend against generic threats. It is often vastly cheaper time-wise compared to creating a valid whitelist - and therefore valuable to many. The catch being that those individuals who would spend the time on a whitelisting likely already have well developed web applications.  However that being said, the majority of web applications I've come across, I would consider 'poorly designed' (personal opinion). As has been seen in the news even websites developed by experienced developers at well established organizations often exhibit vulnerabilities. Sometimes these are even due to using third party packages/tools. In fact things like markdown typically come as a prebuilt package and although extremely useful for the end user open a lot of attack surface from a security perspective, as practitioners of the field if given the opportunity I would much rather choose a more restrictive model but again this is always a trade off between usability and security.

Hopefully this more clearly explains our viewpoint for writing these rules. 


<rant>
Again, I must remind you that tools like ModSecurity are designed to be part of a defense in depth and not a end all solution. A result of this design allows individuals to turn off problematic protections in select areas (such as your false positives) while still being overall secure. To look at modsec as silver bullet is simply naive. Its actually frustrating how many times I've heard some people advise against WAFs on the basis that they don't solve all your problems. This is of course true of all defenses including in depth, line by line, code review. 
</rant>
